algorithms.tex

```diff
@@ -1412,6 +1412,36 @@
 behave as if they are implemented by calling \tcode{begin} and \tcode{end} on
 the \tcode{Range} and dispatching to the overload that takes separate
 iterator and sentinel arguments.
+
+\pnum
+\newtxt{Some algorithms declare both an overload that takes a \tcode{Range}
+and an \tcode{Iterator}, and an overload that takes two \tcode{Range}
+arguments. Since an array type~(\cxxref{basic.compound}) both satisifies
+\tcode{Range} and decays to a pointer~(\cxxref{conv.array}) which satisfies
+\tcode{Iterator}, such overloads are ambiguous when an array is passed as
+the second argument. Implementations shall provide a mechanism to resolve
+this ambiguity in favor of the overload that takes two ranges.}
+
+\pnum
+\newtxt{The number and order of template parameters for algorithm declarations
+is unspecified, except where explicitly stated otherwise. The nominal parameters
+specified herein for each function template declaration are only significant in
+that they specify the association of template constraints with function
+parameters.}
+
+\pnum
+\newtxt{Despite that the algorithm declarations nominally accept parameters
+by value, it is unspecified when and if the argument expressions are used to
+initialize the actual parameters except that any such initialization shall be
+sequenced before~(\cxxref{intro.execution}) the algorithm returns.
+\enternote The behavior of a program that modifies the values of the actual
+argument expressions is consequently undefined unless the algorithm return
+happens before~(\cxxref{intro.multithread}) any such modifications. \exitnote}
+
+\ednote{The intent is to allow algorithm implementations to accept parameters
+by forwarding reference and delay or omit initialization of the actual
+parameter, either as an optimization or to implement the array disambiguation
+requirement described above.}
```

Comment:
Every requirement in the specification implicitly introduces UB, at least this one spells it out clearly. The algorithm will probably initialize new objects from your expressions, but it may call back into your code before and/or after doing so, so don't do anything stupid. Otherwise something as simple as:
```c++
bool any_of(Iterator i, Sentinel<Iterator> s, XXX&& pred_, YYY&& proj_) {
  if (i != s) {
    auto pred = foo(pred_);
    auto proj = foo(proj_);
    // ...
  }
  return true;
}
```
isn't possible under as-if, since (a) the call of a possibly user-supplied `!=` happens before pred and proj are initialized, and (b) pred and proj may not be initialized *at all*, which is definitely observable.